### ISSUE TITLE
AndroidManifest.xml and assets not found for RobolectricGradleTestRunner in Android Studio

### ISSUE THREAD

**Issue by feantury** [robolectric/robolectric#1896]
I am trying to create a test for my Application class, so I set it in the config like:
```java
@RunWith(RobolectricGradleTestRunner.class)
@Config(application = MyApplication.class, constants = BuildConfig.class)
```
The problem comes on how to run them in the Android Studio. We have two options:
- Using a gradle task (needed for CI).
- Using the jUnit built-in test runner (day a day testing in the IDE).

I want both to work in my project but there is a problem that comes up always. When I run it from Android Studio jUnit built-in system, Intellij sets the working directory to my root project, not the module. Of course we can change it to $MODULE_DIR$ but there is a well known issue that makes all tests to pass even if they have failed and the debugger is not allowed when this variable is set, so not an option.

So the situation is:
- We run it from Android Studio jUnit -> Working directory is the root project.
- We run it from Gradle -> Working directory depends where you execute it.

Since the root can change, the manifest location is not found for both situations as it is static in @Config and not controlled in RobolectricGradleTestRunner.

Beside that, assets folder is not found it we place it in a test folder because RobolectricGradleTestRunner tries to find it in the assembled files.

So, to sum up all the problems with this runner:
- Working directory is not consistent and we cannot locate de manifest always in the same dir. Sometimes it is ./moduleName/src/test..., sometimes it is ./src/test...
- Assets located in the test folder cannot be found
- @Config configuration for manifest is ignored for RobolectricGradleTestRunner because it does not call the super implementation

Please let me know if you need more information.

**Comment by jaredsburrows** [comment]
## 1:
`@Config(application = MyApplication.class, constants = BuildConfig.class)`

you should be using:

`@Config(constants = BuildConfig.class)`

## 2:
With this big post, you didn't even share a way for us to reproduce this "error" or "issue". Please share an example test.

## 3:
Focus on getting `Gradle` to work, not the `IDE`.

## 4:
Why are you putting an `assets` in the test folder? Are you changing your app's assets?

**Comment by erd** [comment]
> So, to sum up all the problems with this runner:

The point of `RobolectricGradleTestRunner` is to have reasonable defaults that work for most people. From the description of your requirements, you might be better off creating a custom runner that inherits from `RobolectricTestRunner` and adds your specific logic.

> Of course we can change it to $MODULE_DIR$ but there is a well known issue that makes all tests to pass even if they have failed and the debugger is not allowed when this variable is set

I haven't heard about this, so I'm curious. Is there something you could point me at that describes the problem?

**Comment by sparkym3** [comment]
To be more helpful than the jaredsburrows comments.
1. This is a perfectly valid @Config, but I believe that the application may be overwritten with the application in the manifest if one exists there.

2. This is very clearly stated and doesn't necessarily need a example test.

3. Both gradle and ide should work so that they can be used in different ways. We had similar issue with our projects and found what would work for the gradle run and then modified the default junit working directory to be the path to the submodule, rather than $MODULE_DIR$ (i.e. c:\<project>\<subproject>). Though this is more limiting it does allow us the ease of using the ide.

4. There are many reasons to have pure test assets, as we have them as well. The reason we use this is because we are testing an aar where the asset it relies on is a runtime asset defined by the app that includes the aar.

Also, RobolectricGradleTestRunner should not be calling the super class since all of the parameters are replaced and there would be no gain in calling the super class (other than maybe think about adding it to the cahce, appManifestsByFile).

**Comment by jaredsburrows** [comment]
@sparkym3  "more helpful"? really?

1. Yes it is valid, just unnecessary. All my apps use a custom `Application` and my tests run fine.

2. This is an open source project and without anyway to reproduce the error, it makes it hard for anyone to help. Why do you think @erd asked for more information?

3. Yes, the runner should both work via command line and the IDE, but you can cancel a lot of IDE issues related to testing by confirming your tests *do* work with gradle via command line. He specifies he needs the tests to work with his CI, so you will need to focus on gradle via command line.

4. @sparkym3 and @feantury keep saying "we". Are you guys working together? 

> we use this is because we are testing an aar

Are you testing a `runtime` configuration?

**Comment by sparkym3** [comment]
We, as in my team where I work, as I am not the sole owner of our projects, but no I do not know feantury outside of this post.

It is true that if there is truly an error beyond configuration it would require a test project, but this seems like pure configuration issues and should not require it. erd appears to what more info about why the debugger fails on use of $MODULE_DIR$ rather than the original configuration issue.

The statement of "Focus on getting Gradle to work, not the IDE." is not useful in any way. When you intend for both to work you can work the issue from any direction you like and your personal preference on how to do that is irrelevant. I do agree that I would focus on gradle first, but not all are as comfortable with command line.

The asset I use is a json file that allows for configuration of the aar per build/flavor of the main application. The point is that there are uses for that configuration and it should not be written off.

**Comment by jaredsburrows** [comment]
You are overthinking my statement of "Focus on getting Gradle to work, not the IDE." He explicitly states he needs it for his CI. Of course you need to get Gradle + Robolectric working via command line for the CI, eg: `gradlew testDebug`. I am simply suggesting to solve one problem at a time.

You have a `json` file in your `assets` directory that is read in and decides which `aar` dependency to use?

**Comment by sparkym3** [comment]
No, it always uses the same aar, it just uses different configuration of that aar. I realize it is not an ideal solution and our team is using a non-asset based method within our aars, but we do not have control over that aar as it is a different team's component that we are dependent on.

**Comment by erd** [comment]
>Since the root can change, the manifest location is not found for both situations as it is static in @Config and not controlled in RobolectricGradleTestRunner.

>Beside that, assets folder is not found it we place it in a test folder because RobolectricGradleTestRunner tries to find it in the assembled files.

Given the constraints above, you should create a custom test runner that provides a manifest, resource, and asset path.
